rust/src/interpreter/mod.rs

```diff
@@ -8,4 +6,1260 @@ mod stack;
+                };
+                if revision >= Revision::EVMC_BERLIN
+                    && context.access_storage(addr, &key) == AccessStatus::EVMC_ACCESS_COLD
+                {
+                    dyn_gas += 2100;
+                }
+                consume_gas(&mut gas_left, dyn_gas)?;
+                gas_refund += gas_refund_change;
+                code_state.next();
+            }
+            Opcode::Jump => {
+                consume_gas(&mut gas_left, 8)?;
+                let [dest] = stack.pop()?;
+                code_state.try_jump(dest)?;
+            }
+            Opcode::JumpI => {
+                consume_gas(&mut gas_left, 10)?;
+                let [dest, cond] = stack.pop()?;
+                if cond == u256::ZERO {
+                    code_state.next();
+                } else {
+                    code_state.try_jump(dest)?;
+                }
+            }
+            Opcode::Pc => {
+                consume_gas(&mut gas_left, 2)?;
+                stack.push(code_state.pc())?;
+                code_state.next();
+            }
+            Opcode::MSize => {
+                consume_gas(&mut gas_left, 2)?;
+                stack.push(memory.len())?;
+                code_state.next();
+            }
+            Opcode::Gas => {
+                consume_gas(&mut gas_left, 2)?;
+                stack.push(gas_left)?;
+                code_state.next();
+            }
+            Opcode::JumpDest => {
+                consume_gas(&mut gas_left, 1)?;
+                code_state.next();
+            }
+            Opcode::TLoad => {
+                check_min_revision(Revision::EVMC_CANCUN, revision)?;
+                consume_gas(&mut gas_left, 100)?;
+                let [key] = stack.pop()?;
+                let addr = message.recipient();
+                let value = context.get_transient_storage(addr, &key.into());
+                stack.push(value)?;
+                code_state.next();
+            }
+            Opcode::TStore => {
+                check_min_revision(Revision::EVMC_CANCUN, revision)?;
+                check_not_read_only(message, revision)?;
+                consume_gas(&mut gas_left, 100)?;
+                let [key, value] = stack.pop()?;
+                let addr = message.recipient();
+                context.set_transient_storage(addr, &key.into(), &value.into());
+                code_state.next();
+            }
+            Opcode::MCopy => {
+                check_min_revision(Revision::EVMC_CANCUN, revision)?;
+                consume_gas(&mut gas_left, 3)?;
+                let [dest_offset, offset, len] = stack.pop()?;
+                if len != u256::ZERO {
+                    memory.copy_within(offset, dest_offset, len, &mut gas_left)?;
+                }
+                code_state.next();
+            }
+            Opcode::Push1 => push(1, &mut code_state, &mut stack, &mut gas_left)?,
```

Comment:
`log` needs to be generic because it calls `Stack::pop` which needs the const generic parameter `N`.

Reply:
I see. Not sure about code cache concerns, it seems like a premature performance concern to me, wold require heavy benchmarking to be justified.
 At this phase, I would focus on coherence and homogeneity of the code to facilitate reading.

Reply:
I can see how this would be beneficial for homegeneity but I still don't like the idea of making a function generic if it provides (aside from homogeneity) no advantages. This was not done to improve performance but because to me it simply feels wrong to make them generic. I guess this is a matter of taste to some degree. And it's not that push, dup and swap are not generic like log but more that log is generic because it needs to be.

Reply:
After thinking about it a little bit more, I think the generated machine code is probably the same for both versions anyway, because the functions are so small that they are likely to get inlined. If that happens it no longer matters if we have a const generict or a const as an argument.
This means it is simply a matter of taste and readability. If you think it improves readability a lot I can make them generic. Otherwise I would be in favor of leaving them non-generic.

Reply:
To me, it was just odd to find some generic and non-generic functions, based in the arity of operators. 
I leave this to Herbert,